services/geocoding.py

```diff
@@ -70,4 +76,41 @@ def enhance(self, entries: List[PointOfInterest]) -> List[PointOfInterest]:
                 entry.lat = point.lat
                 entry.lng = point.lng
 
+            if not entry.lat or not entry.lng:
+                continue
+            if entry.country and entry.city and entry.address:
+                continue
+            self.enhance_by_reverse_lookup(entry)
+
         return entries
+
+    def enhance_by_reverse_lookup(self, entry: PointOfInterest):
+        try:
+            response = self.make_reverse_geocode_request(
+                entry.lat, entry.lng, self.gmaps)
+        except (googlemaps.exceptions.ApiError, googlemaps.exceptions.Timeout,
+                googlemaps.exceptions.TransportError) as e:
+            self.log(f"Reverse geocode: failed {entry.lat},{entry.lng}: {e}")
+
+        if not response:
+            self.log(f"Reverse geocode: got 0 results {entry.lat},{entry.lng}")
+        geodata = response[0]["address_components"]
+
+        if not entry.country:
+            entry.country = next(
+                r["long_name"] for r in geodata
+                if "country" in r["types"])
+
+        if not entry.city:
+            city_levels = {
+                "administrative_area_level_1",
+                "administrative_area_level_2",
+            }
```

Comment:
why not use `locality` for the city?
![Screenshot 2022-06-12 at 19 42 54](https://user-images.githubusercontent.com/14199/173234062-4b18c08c-ec55-4564-8a63-abfca893b21f.png)

Reply:
Actually I'm not sure what to use, some answers do have locality, some don't, some adm levels 1 or 2 seem appropriate… I guess I'll add all three in order of increasing significance.